File: wallpaper.py
#!/usr/bin/env python

import argparse
import os

class WallPaper(object):
  def __init__(self):
    self._prefix           = os.getenv('HOME') + '/.wallpapers2/'
    self._lock             = self._prefix + 'lock'
    self._hist             = self._prefix + 'history.db'
    self._category         = self._prefix + 'category'
    self._wallpaper_dir    = self._prefix + 'Wallpapers'
    self._current          = self._prefix + 'current'
    self._resolution       = self._prefix + 'resolution'
    self._log              = self._prefix + 'log'
    self._sources          = self._prefix + 'sources'
    self._bgsetter         = 'fbsetbg -a'
    self._default_category = 'all'

    self.category   = self._get_category()
    self.resolution = self._get_resolution()
    self.sources    = self._get_source_dirs()
    self.wallpapers = self._get_wallpapers()

  def __process_args(self):
    if self.args.dump_cache:
      self.dump_cache()
      exit()

    elif self.args.flush_cache:
      self.flush_cache()
      exit()

    elif self.args.clear:
      self.clear()

    elif self.args.unlock:
      self.unlock()
      exit()

    elif self.args.lock:
      self.lock()
      exit()

    elif self.is_locked and not self.args.unlock:
      print "currently locked, bailing out"
      exit()

    elif self.args.previous:
      self.set_previous()
      exit()

  def _get_category(self):
    category = self._default_category

    if self.args.category:
      category = self.args.category
    elif os.path.exists(self._category):
      category = file(self._category).read()

    return category

  def _get_resolution(self):
    resolution = ''

    if self.args.resolution:
      resolution = self.args.resolution
    elif os.path.exists(self._resolution):
      resolution = file(self._resolution).read()

    return resolution


  def _get_source_dirs(self):
    dirs = []

    if self.category == 'all':
      for source in file(self._sources).read():
        dirs.append(self._wallpaper_dir + '/' + source)

    else

    return dirs

  def _get_wallpapers(self):
    wallpapers = []

    for source in self.sources:
      for root, dirs, files in os.walk(source):
        for file in files
          wallpapers.append(os.path.join(root,file))

    return wallpapers

  def parse_args(self):
    parser = argparse.ArgumentParser(description='Wallpaper changer thingus')

    parser.add_argument('-c',
        '--category',
        help='Wallpaper Category',
        action='store_true')

    parser.add_argument('-r',
        '--resolution',
        help='Wallpaper resolution',
        action='store_true')

    parser.add_argument('-d',
        '--dump-cache',
        help='Display wallpaper cache',
        action='store_true')

    parser.add_argument('-f',
        '--flush-cache',
        help='Flush the wallpaper cache',
        action='store_true')

    parser.add_argument('-l',
        '--lock',
        help='Lock the current wallpaper',
        action='store_true')
    
    parser.add_argument('-u',
        '--unlock',
        help='Unlock the curent wallpaper',
        action='store_true')

    parser.add_argument('-C',
        '--clear',
        help='Clear previous category and resolution',
        action='store_true')

    parser.add_argument('-p',
        '--previous',
        help='Set wallpaper to the previous wallpaper',
        action='store_true')

    self.args = parser.parse_args()
    self.__process_args()

  def lock(self):
    f = open(self._lock,'w')
    f.close()

  def unlock(self):
    if self.is_locked():
      os.remove(self._lock)

  def is_locked(self):
    return os.path.exists(self._lock)

  def set_previous(self):
    print "set_previous(self)"

  def clear(self):
    print "clear(self)"

  def flush_cache(self):
    print "flush_cache(self)"

  def dump_cache(self):
    print "dump_cache(self)"

if __name__ == '__main__':
  wallpaper = WallPaper()
  wallpaper.parse_args()
  wallpaper.set_wallpaper()

File: wallpaper.pl
#!/usr/bin/env perl

use local::lib;
use Modern::Perl;
use DB_File;
use Data::Dumper;
use File::Find::Object;
use Getopt::Long qw(HelpMessage);
use File::Spec;
use File::Stat;
use IO::All;
use DateTime;

my $PREFIX           = qq($ENV{HOME}/.wallpapers);
my $LOCK             = qq{$PREFIX/lock};
my $CATEGORY         = qq{$PREFIX/category};
my $WALLPAPER_DIR    = qq{$PREFIX/Wallpapers};
my $CURRENT          = qq{$PREFIX/current};
my $PREVIOUS         = qq{$PREFIX/previous};
my $SOURCES          = qq{$PREFIX/sources};
my $DEFAULT_CATEGORY = q{all};

unless (-e $WALLPAPER_DIR) {
  die "Wallpaper directory, $WALLPAPER_DIR, does not exist";
}

my %history = ();
tie %history, 'DB_File', qq{$PREFIX/history};    ## no critic (ProhibitTies)

GetOptions(
  'category|c=s'  => sub {$_[1] > io($CATEGORY)},
  'clear'         => sub {unlink $CATEGORY},
  'flush-cache|f' => sub {
    %history = ();
    exit;
  },
  'lock|l'        => sub {2 > io($LOCK)},
  'unlock|u'      => sub {unlink($LOCK)},
  'previous|p'    => sub {
    set_wallpaper(io($PREVIOUS)->all);
    exit;
  },
  'image|i=s' => sub {
    set_wallpaper($_[1]);
    exit;
  },
  'dump-cache|d' => sub {
    print {*STDOUT} Dumper \%history;
    exit;
  },
  'h|help' => \&HelpMessage,
  )
  or HelpMessage(1);

exit if -e $LOCK;

set() unless -e $LOCK;

sub set {
  my @wallpaper_dirs = get_wallpaper_dirs();
  my @wallpapers     = get_wallpapers(@wallpaper_dirs);
  return _set(\@wallpapers);
}

sub _set {
  my ($papers)  = @_;
  my $rc        = 0;
  my $set_paper = 0;

  if (scalar @{$papers} == 1) {
    $rc        = set_wallpaper($papers->[0]);
    $set_paper = 1;
  } else {
    while (my $paper = get_random_wallpaper($papers)) {
      next if exists $history{$paper};
      $rc        = set_wallpaper($paper);
      $set_paper = 1;
      last;
    }
  }

  if (not $set_paper) {
    %history = ();
  } else {
    return $rc;
  }

  return _set($papers);
}

sub get_wallpaper_dirs {
  my @paths = ();
  my $category = (-e $CATEGORY) ? io($CATEGORY)->getline : $DEFAULT_CATEGORY;

  if ($category eq $DEFAULT_CATEGORY) {

    return map {File::Spec->join($WALLPAPER_DIR, $_)} io($SOURCES)->chomp->getlines;
  } else {
    push @paths, $WALLPAPER_DIR;
  }

  push @paths, $category;

  my $dir = File::Spec->join(@paths);
  die qq{Wallpaper directory ($dir) does not exist} if not -e $dir;

  return ($dir);
}

sub get_wallpapers {
  my (@dirs) = @_;

  my $now     = DateTime->now();
  my $tree    = File::Find::Object->new({}, @dirs);
  my @papers  = ();
  my %weights = (
    86400   => 1000,
    604800  => 500,
    2592000 => 200,
  );

  while (my $leaf = $tree->next()) {
    next if -d $leaf;
    next if exists $history{$leaf};

    my $stat = File::Stat->new($leaf);

    for my $weight (keys %weights) {
      if ((($now->epoch - $weight) < $stat->ctime)) {
        push @papers, $leaf for (1 .. ($weights{$weight} - 1));
      }
    }

    push @papers, $leaf;
  }

  return @papers;
}

sub get_random_wallpaper {
  my ($papers) = @_;

  my $pos   = int(rand(scalar @{$papers}));
  my $paper = $papers->[$pos];

  return splice(@{$papers}, $pos, 1);
}

sub set_wallpaper {
  my ($paper) = @_;

  my $rv = io->pipe(qq{fbsetbg -f '$paper'})->all;

  $history{$paper} = 1;
  rename $CURRENT, $PREVIOUS;
  shift > io($CURRENT);

  return $rv;
}

__END__

=head1 NAME

wallpaper.pl - Script to manage my desktop wallpapers.

=head1 SYNOPSIS

wallpaper.pl [OPTIONS]

  Options:

    -c, --category    Wallpaper category
    -f, --flush-cache Flush the wallpaper cache
    -d, --dump-cache  Dump the current wallpaper cache to STDOUT
    -l, --lock        Lock the current wallpaper
    -u, --unlock      Unlock the current wallpaper
        --clear       Clear the previous wallpaper category
    -p, --previous    Set the wallpaper to the previous paper
    -i, --image       Set image as the current wallpaper

=head1 DESCRIPTION

This sciprt is used to manage the desktop wallpaper displayed on my system.

=head1 ENVIRONMENT VARIABLES

None

=head1 VERSION

0.2

=cut
